Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -738,27 +738,35 @@ internal extension ByteBuffer {
         self.write(staticString: crlf)
     }
 }
-extension HTTPHeaders: Sequence {
-    public typealias Element = (name: String, value: String)
-
-    /// An iterator of HTTP header fields.
-    ///
-    /// This iterator will return each value for a given header name separately. That
-    /// means that `name` is not guaranteed to be unique in a given block of headers.
-    public struct Iterator: IteratorProtocol {
-        private var headerParts: Array<(String, String)>.Iterator
-
-        fileprivate init(headerParts: Array<(String, String)>.Iterator) {
-            self.headerParts = headerParts
-        }
 
-        public mutating func next() -> Element? {
-            return headerParts.next()
+extension HTTPHeaders: RandomAccessCollection {
+    public typealias Element = (name: String, value: String)
+    public struct Index: Comparable {
+        let base: Int
+        public static func < (lhs: Index, rhs: Index) -> Bool {
+            return lhs.base < rhs.base
         }
     }
-
-    public func makeIterator() -> Iterator {
-        return Iterator(headerParts: headers.map { (self.string(idx: $0.name), self.string(idx: $0.value)) }.makeIterator())
+    
+    public subscript(position: Index) -> (name: String, value: String) {
+        let header = self.headers[position.base]
+        return (name: self.string(idx: header.name), value: self.string(idx: header.value))
+    }
+    
+    public var startIndex: Index {
+        return Index(base: self.headers.startIndex)
+    }
+    
+    public var endIndex: Index {
+        return Index(base: self.headers.endIndex)
+    }
+    
+    public func index(before i: Index) -> Index {
+        return Index(base: i.base - 1)
+    }
+    
+    public func index(after i: Index) -> Index {
+        return Index(base: i.base + 1)
```

Comment:
So there's nothing wrong with these functions, but I'd like to make them a bit more future-proof.

Right now what we're doing is delegating to the index of our backing storage, which is `Array<HTTPHeader>`. I think we may be best served by explicitly stating this, and then moving to use the correct functions for manipulating those indices.

To do this, we can redefine `Index` such that `base` is of type `Array<HTTPHeaders>.Index`. This is a no-op change, as that's just a typealias for `Int`, but it more correctly indicates the intended design (and makes us resilient against changing the backing store).

Once we've done that, we can change `index(before:)` and `index(after:)` to just call those functions on `Array`, rather than hard-code knowledge about how `Array`'s indexing works.

Does that make sense?

Reply:
Yes, that's a nice one. Just pushed that update.